### PR TITLE
[Collective] Some necessary abstraction of collective calls before introducing stream management

### PR DESCRIPTION
**Pull request by zhisbug**
See a list below:

1.  PR #12637 ( **merged**) ~~Basic infrastructure; an in-actor collective interface `ray.util.collective.init_collective_group(*args, **kwargs)`; support for two collectives `allreduce` and `barrier`; some testing infrastructure, etc.~~
2. (PR #12874 **pending merge**) Driver-program interface, which includes: (1) the second interface: `actor.options(collective_options, ...).remote()` and the third interface `declare_collective_group(actors, collective_options, ...)`.
3. (PR #12864 **merged**) ~~Support for other collectives: `allgather`, `broadcast`, `reduce`, `reducescatter`; refactor the tests into distributed tests and single-node cluster tests~~. 
3.5 (PR #12935 **merged**) ~~send/recv P2P communication apis.~~
4. Communicator caching,  and support for num_gpus > 2 per actor/task.
5. (**first part of this one**) CUDA stream management.
6. docs, examples, etc.

## Related issue number
See this [rfc-collective-in-ray](https://github.com/ray-project/RFC/blob/main/rfc-20201119-collective-in-ray/20201119-collective-in-ray.md)

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zhisbug** [comment]
@richardliaw : rebase done; this one is ready to merge.

It contains some minor improvement of collective calls before we merge the next [bigger PR#13030](https://github.com/ray-project/ray/pull/13030)

File: python/ray/util/collective/collective.py [hunk 1]
```diff
@@ -156,17 +156,20 @@ def declare_collective_group(actors,
         raise RuntimeError("Each actor should correspond to one rank.")
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:156]: can you also provide more info here (i.e. what ranks were received)

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 2]
```diff
@@ -422,10 +431,51 @@ def _get_cuda_stream():
         # TODO: implement a simple stream manager.
         return cupy.cuda.Stream.null
 
-    # Note(Hao): too many bipolate code -- make some abstraction.
-    # def _collective_call(self, *args):
-    #     """Private method to encapsulate all collective calls"""
-    #     pass
+    def _collective(self, input_tensor, output_tensor, collective_fn,
+                    preprocess_fn, postprocess_fn):
```
Comment (richardliaw) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:435]: can you make these named arguments and default preprocess_fn and postprocess_fn to None?

Then, you don't need to drop the lambda everywhere

Comment (zhisbug) [reply python/ray/util/collective/collective_group/nccl_collective_group.py:435]: sounds perfect!

**Review by richardliaw — APPROVED** [review]
small nits, ping me when addressed

**Comment by zhisbug** [comment]
@richardliaw addressed. Merge?

**Review by richardliaw — APPROVED** [review]

**Pull request merged**